vehicle/src/Vehicle/Backend/LossFunction/Compile.hs

```diff
@@ -217,6 +197,203 @@ compileQuant :: V.Quantifier -> Quantifier
 compileQuant V.Forall = All
 compileQuant V.Exists = Any
 
+
+----------------------------------------------------------------------------------------------------
+--handling normalisation options
+
+
+normalisationOptions :: NormalisationOptions
+normalisationOptions = Options
+  { declContext                 = mempty
+  , boundContext                = mempty
+  , normaliseDeclApplications   = False
+  , normaliseLambdaApplications = False
+  , normaliseStdLibApplications = False
+  , normaliseBuiltin            = normBuiltin
+  , normaliseWeakly             = False
+  }
+
+
+normBuiltin :: V.Builtin -> Bool
+normBuiltin b = case b of
+  V.TypeClassOp t -> case t of
+    V.FromNatTC {} -> True
+    V.FromRatTC    -> True
+    V.FromVecTC {} -> True
+    V.NotTC        -> True
+    V.AndTC        -> True
+    V.OrTC         -> True
+    V.ImpliesTC    -> True
+    V.MapTC        -> True
+    V.NegTC        -> True
+    V.AddTC        -> True
+    V.SubTC        -> True
+    V.MulTC        -> True
+    V.DivTC        -> True
+
+    _              -> False
+
+  V.FromNat {}       -> True
+  V.FromRat {}       -> True
+  V.FromVec {}       -> True
+  V.Foreach{}        -> True
+
+  _                  -> False
+
+data DifferentialLogicImplementation = DifferentialLogicImplementation
+    { compileAnd          :: LExpr -> LExpr -> LExpr
+    , compileOr           :: LExpr -> LExpr -> LExpr
+    , compileNot          :: LExpr -> LExpr
+    , compileImplication  :: LExpr -> LExpr -> LExpr
+
+    , compileLe           :: LExpr -> LExpr -> LExpr
+    , compileLt           :: LExpr -> LExpr -> LExpr
+    , compileGe           :: LExpr -> LExpr -> LExpr
+    , compileGt           :: LExpr -> LExpr -> LExpr
+    , compileEq           :: LExpr -> LExpr -> LExpr
+    , compileNeq          :: LExpr -> LExpr -> LExpr
+
+    , compileBool         :: Bool -> Double
+    }
+
+
+--------------------------------------------------------------------------------
+-- different avilable differentiable logics (types of translation from the constraint to loss function) are
+-- DL2
+-- Godel
+-- Lukasiewicz
+-- Product based
+-- Yager
+
+--they can be found in Vehicle.Backend.Prelude and the default option if none is provided is DL2.
+
+-- part of the syntax translation that differ depending on chosen DL are:
+    -- logical connectives (not, and, or, implies)
+    -- comparisons (<, <=, >, >=, =, !=)
+
+dl2Translation :: DifferentialLogicImplementation
```

Comment:
Oh and it might be nice to document which paper you took each translation from in a comment on the translation?